### ISSUE TITLE
Implement the OAuth flow for the Setup CTA Banner

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#8132]
## Feature Description

Implement the redirection to OAuth to grant the _edit_ scope for the Setup CTA Banner.

Note that this issue only covers the happy path, the unhappy path will be implemented separately via https://github.com/google/site-kit-wp/issues/8134.

See [setup CTA banner > setup logic](https://docs.google.com/document/d/1MGD5Djy6AeeZC4zBtHqS-lQEWD9jw0kf-IIWw-jLCFU/edit#heading=h.wuovnzlsox34) in the design doc.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* If the user does not have the Analytics Admin Edit scope, then clicking on `Enable groups` in the Setup CTA should trigger the standard OAuth flow used in the plugin.
    * On successful completion, the user should be redirected back to the dashboard where the rest of the happy path implemented in #8131 should be followed.

## Implementation Brief

* [ ] Update `assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceSegmentationSetupCTAWidget.js`
  * [ ]  Add a callback function for `Enable groups` CTA:
    * Check if user does not has Analytics 4 Edit scope, to trigger the `setPermissionScopeError` for `ANALYTICS_EDIT_SCOPE`. You can see an example [here](https://github.com/google/site-kit-wp/blob/5913a7fde140271d2ad64b02a5c34505d97e59a2/assets/js/components/KeyMetrics/MetricsSelectionPanel/Footer.js#L189-L203) 
    * Append query argument `notification` with value, say `audience_segmentation`, to bypass `authentication_success` banner. You can see an example [here](https://github.com/google/site-kit-wp/blob/5913a7fde140271d2ad64b02a5c34505d97e59a2/assets/js/modules/analytics-4/components/dashboard/EnhancedMeasurementActivationBanner/SetupBanner.js#L73-L75)
    * Set `autoSubmit` value on `CORE_FORMS`, by dispatching `setValues` action, and assign it to, say, `AUDIENCE_SEGMENTATION_SETUP_FORM`. You can see an example [here](https://github.com/google/site-kit-wp/blob/efdd75e4de73791a2e09eba73b2c4b6f57789304/assets/js/modules/analytics-4/components/common/AccountCreate/index.js#L151)
* [ ] Update the setup component implemented in #8131 to proceed with the setup step (when user has the `ANALYTICS_EDIT_SCOPE`) if query argument `autoSubmit` is `true`

### Test Coverage

* Update jest tests to verify this scenario of jumping to the setup step after OAuth redirection. You can see example of mocking the redirect url [here](https://github.com/google/site-kit-wp/blob/5913a7fde140271d2ad64b02a5c34505d97e59a2/assets/js/components/notifications/BannerNotifications.test.js#L98-L114)

## QA Brief

1. Setup Site Kit with the `audienceSegmentation` feature flag enabled.
2. Connect an Analytics property which is not in the "gathering data" state - preferably a property which already has enhanced measurement enabled, so it's not necessary to grant the _edit_ scope during setup.
3. If it _was_ necessary to grant the _edit_ scope during setup, disconnect and then reconnect Site Kit to clear the granted scope.
4. Refresh the dashboard, as the Audience Segmentation Setup CTA won't show up on the first page load while the gathering data state is being resolved.
5. Click on the **Enable groups** button - this should navigate to the OAuth flow. Proceed through the OAuth flow and grant the requested permissions, upon which the browser should navigate back to the Site Kit dashboard. The CTA should now be in its loading state with the text "Enabling groups" and a spinner showing in the button.
6. Once the operation is completed, the CTA banner will disappear and a success notice will be displayed with the heading:

```text
Success! Visitor groups added to your dashboard
```

## Changelog entry

* Navigate to the OAuth flow as needed when setting up Audience Segmentation from the Setup CTA Banner.

**Comment by tofumatt** [comment]
ACs sound good 👍🏻

**Comment by techanvil** [comment]
Hi @zutigrm, the IB is looking good, however I am not sure we should use the `notification` query param as the condition for reshowing the banner and continuing to the next step.

It's not an approach we have taken to date, rather we have a pattern of setting an `autoSubmit` form value, which is persisted via the store snapshot mechanism and restored on the next page load. For example:

https://github.com/google/site-kit-wp/blob/efdd75e4de73791a2e09eba73b2c4b6f57789304/assets/js/modules/analytics-4/components/common/AccountCreate/index.js#L151-L163

We then use the `autoSubmit` value in the conditions for restoring the previous UI state and continuing.

I'd suggest we take the same approach here, in part for consistency, but also because it will be useful for the unhappy path too. We can use it for showing the error modal - the `notification` param may not be present in this scenario.

**Comment by zutigrm** [comment]
@techanvil Thank you for the feedback, I forgot about `autoSubmit` 🤦 . I updated the IB

**Comment by techanvil** [comment]
Thanks @zutigrm! IB LGTM. :white_check_mark:

**Comment by ankitrox** [comment]
Adding "Blocked by" #8707 because some logic inside the component `AudienceSegmentationSetupCTAWidget` will get changed in 8707 because of which tests for this issue needs to accommodate the change.

**Comment by techanvil** [comment]
Hey @ankitrox, I've left a few comments on the PR.

With regard to the QAB - although it could be enough to give the QA team sufficient direction, it would be useful to provide a bit more detail.

I've gone ahead and rewritten it a bit myself, please review the changes and make sure you are happy with them. Note that I've completely omitted the point about visiting https://myaccount.google.com/connections to remove permissions - we don't generally need to do that, in the scenario here simply disconnecting and reconnecting Site Kit is sufficient.

**Comment by ankitrox** [comment]
Hi @techanvil ,

We can discuss about the use of mocks for the actions later when @aaemnnosttv is available. However; I have changed the tests to make use of assertions for `enableAudienceGroups` action so that those are more of the integration in nature as per the convention. 

Please note that I have made the use of `muteFetch` for reports endpoint as this is not really specific to the tests being performed and to keep the tests readable.

Also, updated these tests so that this ticket can be moved ahead because there are some other tickets which are dependent on this one.

Assigning this to you for re-review.

Thank you

**Comment by techanvil** [comment]
Thanks @ankitrox! As per my review comment, I am not seeing the updated version of the tests. Maybe you need to push a recent change?

**Comment by ankitrox** [comment]
@techanvil Sorry, that was my bad! I committed the changes, but missed to push them.

I have pushed the changes and those should be available. Please refer to this [commit](https://github.com/google/site-kit-wp/pull/8741/commits/b5ac326f219ffb16c61331874da94705c0894961) where I have removed mock's implementation just to stick to our convention.

**Comment by techanvil** [comment]
No worries, thanks @ankitrox! As per my latest review this needs one more pass, please take a look.

**Comment by ankitrox** [comment]
@techanvil Thank you.

Tests updated as per suggestions.

**Comment by kelvinballoo** [comment]
### QA Update ❌ 

I have tried to test this but it's not working as expected.
After clicking on 'Enable groups', I get redirected to the OAuth flow but once I am back to the dashboard, the success message appears after a while but the enable group still has the spinning loading icon.

Video is attached below for reference:

<Details>
https://github.com/google/site-kit-wp/assets/125576926/9ddf036b-77ca-4980-bca8-74bcce2126b3
</Details>

**Comment by ankitrox** [comment]
Thanks @kelvinballoo ,

As per our [slack conversation](https://10up.slack.com/archives/C077DB69US2/p1718975941739479), we need the analytics property with write access (in order to create audiences in this case).

Can you please test it according to that and let me know if there is something you want me to do in this case?

Thanks

**Comment by kelvinballoo** [comment]
### QA Update ⚠️ 
Noted on that @ankitrox 

I have tested this happy flow with an analytics property with write access and it's looking good. Refer to below for the desktop flow. ✅ 
<Details>

https://github.com/google/site-kit-wp/assets/125576926/e287f678-1673-4c44-93c3-06ce3a473331


</Details>

It's worth noting that the success message box design on mobile doesn't align with Figma. @ankitrox , let me know if this needs another ticket or there is an existing ticket as it's most likely out of scope for this ticket. ⚠️ 

<Details>
Mobile implementation (iPhone 15prox max):
<img width="325" alt="Screenshot 2024-06-25 at 12 29 02" src="https://github.com/google/site-kit-wp/assets/125576926/c4510cec-0b8b-45f9-948d-849f6b2d71b3">

Figma: 
<img width="339" alt="Screenshot 2024-06-25 at 12 28 06" src="https://github.com/google/site-kit-wp/assets/125576926/07e183d8-fa4e-44c8-ab62-ac8a325396a8">

</Details>

**Comment by kelvinballoo** [comment]
### QA Update ✅ 

After latest conversations, the mobile issue is not part of this ticket. Hence, another ticket has been raised here: https://github.com/google/site-kit-wp/issues/8932

As for this ticket, it's good to go as the happy path for the setup is working properly.
Video is attached for reference below.

Moving this ticket to Approval.

<Details>

https://github.com/google/site-kit-wp/assets/125576926/e287f678-1673-4c44-93c3-06ce3a473331

</Details>

**Comment by aaemnnosttv** [comment]
@techanvil we're requesting the edit scope here but that isn't mentioned in the interface, nor do we know for sure that it's needed, do we? We should only request the readonly scope as usual until we actually need the edit scope. When requesting the edit scope we always do so with a explanation as to why. Maybe this is coming in a later issue or was missed during UX but I just wanted to call it out. It's not blocking here.
